### PR TITLE
Add support for Hetzner Private Networks

### LINKED ISSUE #599: Add support for hetzner private network
Hetzner recently added support for private networking, would be good to extend machine-controller to support this.

This is how it looks like with new version of terraform hetzner provider: https://github.com/kubermatic/kubeone/pull/596/files#diff-8aec760bfdbb019e856b85fc10e43205

### PR DESCRIPTION
**Pull request by xmudrii**
**What this PR does / why we need it**:

This PR implements support for Hetzner Private Networks. A new optional field is added to the Hetzner spec called `network` that consumes a network ID or name. It is expected that the network already exists, i.e. that it's created by the user.

To make this work I had to vendor the latest version of the `hcloud-go` library which includes support for networks.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #599 

```release-note
Add support for Hetzner Private Networks
```

### CODE REVIEW

**Comment by xmudrii** [comment]
@alvaroaleman When you get some time please take a look at questions I left in the PR description. It's nothing urgent, so no hurry. :slightly_smiling_face:

**Comment by alvaroaleman** [comment]
> Running machine-controller binary on the local machine complains about plugins.

That is supposed to work :thinking: 

> Should I add some E2E (or other) tests for this feature?

Hmm, I think this is a bit tricky, because something has to create the network we use there beforehand

> Should we support just a single network or multiple networks?

I guess there is no harm in supporting more? I don't really see the usecase, but if we don't, someone will complain sooner or later I imagine

> How machine-controller knows which network interface to use by default,

I don't understand the question, where does the machine-controller care about the network interface?

**Comment by xmudrii** [comment]
@alvaroaleman 
> I don't understand the question, where does the machine-controller care about the network interface?

That's what I'm asking. :slightly_smiling_face: Does the machine-controller care about the network interfaces anywhere? As far as I get all the existing code, it's just enough to attach the machine to the network and that should be it.

Also, have I broke something? I guess this is not really a flake. :sweat_smile:

**Comment by alvaroaleman** [comment]
> Also, have I broke something? I guess this is not really a flake. ￼

not sure what was going on with the tests yesterday, probably hetzner running out of capacity.

>  Does the machine-controller care about the network interfaces anywhere? 

Not really, it uses the IP to match the machine to its node on platforms that do not have a cloudprovider like Hetzner but thats about it

File: pkg/cloudprovider/provider/hetzner/provider.go [hunk 1]
```diff
@@ -386,6 +414,9 @@ func (s *hetznerServer) Addresses() []string {
 	for _, fips := range s.server.PublicNet.FloatingIPs {
 		addresses = append(addresses, fips.IP.String())
 	}
+	for _, privateNetwork := range s.server.PrivateNet {
+		addresses = append(addresses, privateNetwork.IP.String())
```
Comment (xmudrii) [inline pkg/cloudprovider/provider/hetzner/provider.go]: Is there a correct order here, e.g. should private addresses go before or after other addresses?

**Comment by xmudrii** [comment]
> Not really, it uses the IP to match the machine to its node on platforms that do not have a cloudprovider like Hetzner but thats about it

Okay, that makes sense. I think this PR has everything covered now and should be ready for review. :slightly_smiling_face:

File: pkg/cloudprovider/provider/hetzner/provider.go (hunk 1 above)
Comment (alvaroaleman) [reply pkg/cloudprovider/provider/hetzner/provider.go]: Nom, doesn't matter

File: examples/hetzner-machinedeployment.yaml [hunk 2]
```diff
@@ -47,6 +47,9 @@ spec:
             # Optional
             datacenter: ""
             location: "fsn1"
+            # Optional: network IDs or names
+            networks:
+              - ""
```
Comment (alvaroaleman) [inline examples/hetzner-machinedeployment.yaml]: can we put something like `<<my-network>>` between the braces?

File: pkg/cloudprovider/provider/hetzner/provider.go [hunk 3]
```diff
@@ -152,6 +161,14 @@ func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
 		}
 	}
 
+	if len(c.Networks) != 0 {
+		for _, network := range c.Networks {
+			if _, _, err = client.Network.Get(ctx, network); err != nil {
+				return fmt.Errorf("failed to get network: %v", err)
```
Comment (alvaroaleman) [inline pkg/cloudprovider/provider/hetzner/provider.go]: nit: Can we put the name of the network we couldnt get into the error?

File: examples/hetzner-machinedeployment.yaml (hunk 2 above)
Comment (xmudrii) [reply examples/hetzner-machinedeployment.yaml]: Added `<< YOUR_NETWORK >>`.

File: pkg/cloudprovider/provider/hetzner/provider.go (hunk 3 above)
Comment (xmudrii) [reply pkg/cloudprovider/provider/hetzner/provider.go]: Fixed

**Comment by xmudrii** [comment]
@alvaroaleman Fixed, PTAL

**Pull request merged**